### PR TITLE
Fix error swallowing pipeline async iterators

### PR DESCRIPTION
**Pull request by ronag**
##### Checklist

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

### CODE REVIEW

File: lib/internal/streams/pipeline.js [hunk 1]
```diff
@@ -125,7 +126,9 @@ async function pump(iterable, writable, finish) {
     }
     writable.end();
   } catch (err) {
-    finish(err);
+    error = err;
```
Comment (ronag) [inline lib/internal/streams/pipeline.js:129]: This is a little awkward but we need to ensure finish is always called and only once.

File: lib/internal/streams/pipeline.js [hunk 2]
```diff
@@ -149,36 +152,38 @@ function pipeline(...streams) {
...
-        destroy(error);
-      }
+    if (!error && !final) {
+      return;
+    }
+
+    for (const destroy of destroys) {
+      destroy(error);
     }
+    destroys.length = 0;
```
Comment (mcollina) [inline lib/internal/streams/pipeline.js]: I would prefer if we did a `destroys.shift()` instead.

**Review by mcollina — COMMENTED** [review]
Can you target this PR against mine?

**Comment by ronag** [comment]
> Can you target this PR against mine?

Hmm, GitHub doesn't let me select mcollina/node as a base repository.

File: lib/internal/streams/pipeline.js (hunk 2 above)
Comment (ronag) [reply lib/internal/streams/pipeline.js]: Fixed

**Comment by mcollina** [comment]
I'll pull your changes into mine and add you as co-author then.

**Pull request closed without merging**